udp/udp.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2015 The Chihaya Authors. All rights reserved.
+// Use of this source code is governed by the BSD 2-Clause license,
+// which can be found in the LICENSE file.
+
+// Package udp implements a UDP BitTorrent tracker per BEP 15.
+// IPv6 is currently unsupported as there is no widely-implemented standard.
+package udp
+
+import (
+	"errors"
+	"net"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/pushrax/bufferpool"
+
+	"github.com/chihaya/chihaya/config"
+	"github.com/chihaya/chihaya/tracker"
+)
+
+// Server represents a UDP torrent tracker.
+type Server struct {
+	config  *config.Config
+	tracker *tracker.Tracker
+	done    bool
+	booting chan struct{}
+	sock    *net.UDPConn
+
+	connIDGen *ConnectionIDGenerator
+}
+
+func (s *Server) serve(listenAddr string) error {
+	if s.sock != nil {
+		return errors.New("server already booted")
+	}
+
+	udpAddr, err := net.ResolveUDPAddr("udp", listenAddr)
+	if err != nil {
+		close(s.booting)
+		return err
+	}
+
+	sock, err := net.ListenUDP("udp", udpAddr)
+	defer sock.Close()
+	if err != nil {
+		close(s.booting)
+		return err
+	}
+
+	if s.config.UDPReadBufferSize > 0 {
+		sock.SetReadBuffer(s.config.UDPReadBufferSize)
+	}
+
+	pool := bufferpool.New(1000, 2048)
+	s.sock = sock
+	close(s.booting)
+
+	for !s.done {
+		buffer := pool.TakeSlice()
+		sock.SetReadDeadline(time.Now().Add(time.Second))
+		n, addr, err := sock.ReadFromUDP(buffer)
+
+		if err != nil {
+			if netErr, ok := err.(net.Error); ok && netErr.Temporary() {
+				pool.GiveSlice(buffer)
+				continue
+			}
+			return err
+		}
+
+		start := time.Now()
+
+		go func() {
+			response, action := s.handlePacket(buffer[:n], addr)
+			pool.GiveSlice(buffer)
+
+			if response != nil {
+				sock.WriteToUDP(response, addr)
+			}
+
+			if glog.V(2) {
+				duration := time.Since(start)
+				glog.Infof("[UDP - %9s] %s %s", duration, action, addr)
+			}
+		}()
+	}
+
+	return nil
+}
+
+// Serve runs a UDP server, blocking until the server has shut down.
+func (s *Server) Serve(addr string) {
+	glog.V(0).Info("Starting UDP on ", addr)
+
+	go func() {
+		// Generate a new IV every hour.
+		for range time.Tick(time.Hour) {
+			s.connIDGen.NewIV()
```

Comment:
I'm ok with adding a lock, but here's my reasoning for why this is thread-safe.

In [NewIV](https://github.com/chihaya/chihaya/blob/14a6278de015966e2a81a60387f402e8981eb945/udp/connection.go#L86-L87), `g.iv2` is set before setting `g.iv`, so there is no moment where either variable contains a bad iv (at worst they will be the same).

In [Matches](https://github.com/chihaya/chihaya/blob/14a6278de015966e2a81a60387f402e8981eb945/udp/connection.go#L66-L74), `g.iv` and `g.iv2` are only read once into a temporary local, so there is no opportunity for instruction reordering or anything else to cause them to be read incorrectly.